src/TestCase.ts

```diff
@@ -0,0 +1,173 @@
+    end: serializePosition(range.end),
+  };
+}
+
+export function serializeSelection(selection: Selection): SerializedSelection {
+  return {
+    active: serializePosition(selection.active),
+    anchor: serializePosition(selection.anchor),
+  };
+}
+
+export function serializePosition(position: Position): SerializedPosition {
+  return { line: position.line, character: position.character };
+}
+
+type Command = {
+  actionName: ActionType;
+  partialTargets: PartialTarget[];
+  extraArgs: any[];
+};
+
+type Snapshot = {
+  document: string;
+  clipboard: string;
+  visibleRanges: SerializedRange[];
+  selections: SerializedSelection[];
+};
+
+type DecorationRanges = { [coloredSymbol: string]: SerializedRange };
+
+export type TestCaseFixture = {
+  command: Command;
+  languageId: string;
+  decorations: DecorationRanges;
+  initialState: Snapshot;
+  finalState: Snapshot;
+};
+
+export default class TestCase {
+  command: Command;
+  languageId: string;
+  decorations: DecorationRanges;
+  initialState: Snapshot | null = null;
+  finalState: Snapshot | null = null;
+
+  constructor(
+    command: Command,
+    targets: Target[],
+    navigationMap: NavigationMap
+  ) {
+    const activeEditor = vscode.window.activeTextEditor!;
+
+    this.command = command;
+    this.languageId = activeEditor.document.languageId;
+    this.decorations = this.extractTargetedDecorations(targets, navigationMap);
+  }
+
+  extractPrimitiveTargetKeys(...targets: PrimitiveTarget[]) {
+    const keys: string[] = [];
+    targets.forEach((target) => {
+      if (target.mark.type === "decoratedSymbol") {
+        keys.push(`${target.mark.symbolColor}.${target.mark.character}`);
+      }
+    });
+    return keys;
+  }
+
+  extractTargetKeys(target: Target): string[] {
+    switch (target.type) {
+      case "primitive":
+        return this.extractPrimitiveTargetKeys(target);
+
+      case "list":
+        return target.elements.map(this.extractTargetKeys, this).flat();
+
+      case "range":
+        return this.extractPrimitiveTargetKeys(target.start, target.end);
+
+      default:
+        return [];
+    }
+  }
+
+  extractTargetedDecorations(targets: Target[], navigationMap: NavigationMap) {
+    if (!navigationMap) {
+      return {};
+    }
+
+    const decorationRanges = navigationMap.serializeRanges();
+    const targetedDecorations: DecorationRanges = {};
+    const targetKeys = targets.map(this.extractTargetKeys, this).flat();
+    targetKeys.forEach((key) => {
+      targetedDecorations[key] = decorationRanges[key];
+    });
+    return targetedDecorations;
+  }
+
+  static async getSnapshot(): Promise<Snapshot> {
+    const activeEditor = vscode.window.activeTextEditor!;
+    return {
+      document: activeEditor.document.getText(),
+      selections: activeEditor.selections.map(serializeSelection),
+      visibleRanges: activeEditor.visibleRanges.map(serializeRange),
+      clipboard: await vscode.env.clipboard.readText(),
```

Comment:
I'd be tempted to only bother with the clipboard if it's a copy or paste command